[PM-10439] Vault generator components

nick-livefront opened the pull request:
## 🎟️ Tracking

[PM-10439](https://bitwarden.atlassian.net/browse/PM-10439)

## 📔 Objective

Integrate the `PasswordGeneratorComponent` and `UsernameGeneratorComponent` into the web & browser vaults.
- Added the needed translations for the web
- There are some interesting scenarios occurring with the generation functionality, some of which still appears to be in flight. 

## 📸 Screenshots

|Extension|Web|
|-|-|
|<video src="https://github.com/user-attachments/assets/5ed3da93-0fb4-473e-9f79-45256b6917ca" />|<video src="https://github.com/user-attachments/assets/16f32ac0-f4f7-412f-9c6a-74cb91665514" />

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes


[PM-10439]: https://bitwarden.atlassian.net/browse/PM-10439?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

audreyality commented on libs/tools/generator/components/src/dependencies.ts:
```diff
@@ -43,7 +42,6 @@ const RANDOMIZER = new SafeInjectionToken<Randomizer>("Randomizer");
     InputModule,
     ItemModule,
     JslibModule,
-    JslibServicesModule,
```
Each one of the modules listed in that file I added because it broke something when it wasn't listed. I'd check the generator proper to see whether removing the module causes an error in the generator tab.

That said, I'm less interested on whether or not it fixes the error than why it fixes it. That way I'm unlikely to make the same mistake twice--or it may point to something neither of us are aware of happening during initialization that needs to be changed.

nick-livefront replied:
For sure, from what I could gather was a circular import was occurring. I can't explain why the same structure was working on the browser extension. 

I think the issue stems from the `JslibServicesModule` being re-exported from the `DependenciesModule` as it's imported into other modules directly. 

Correct me if I'm missing something!

audreyality replied:
Hm... that would mean that most of this module's exports are incorrect. Is there a different way I should be sharing the dependencies for the generator components? I suppose I could just add them to each component individually, but wanted to avoid the duplication. (I'm a novice when it comes to Angular.)

nick-livefront replied:
For sure, we're getting to the bounds of my understanding as well 😃 I might rely on @shane-melton here.

My understanding is that the `exports` exposes classes/directives/etc to consumers. Ex, When the `DependencyModule` is imported, than whichever construct imported it would have access to the exports array.

@shane-melton keep me honest

audreyality replied:
Either way, this is resolved. I'll just keep it open so that the answer is logged. :)

shane-melton replied:
> My understanding is that the `exports` exposes classes/directives/etc to consumers. Ex, When the `DependencyModule` is imported, than whichever construct imported it would have access to the exports array.

Yep, if you want to share dependencies `exports` is what you need to populate ([see here](https://v17.angular.io/guide/sharing-ngmodules#sharing-modules)). 

However, I believe `JslibServiceModule` is a bit unique in that it should [only ever be imported at the root `AppModule`](https://v17.angular.io/guide/sharing-ngmodules#sharing-modules) to avoid duplicating the services. By re-exporting here consumers would attempt to import the module again leading to the weird access before ready error (or something along those lines).

audreyality replied:
Ah, I see. Without the export, there are scenarios where the services fail to load. I forget what they are, but it came up during development. I'll be converting the components to module-level components instead of standalone.

In the meantime I'll resolve this so that y'all can merge. My PR will handle importing the module as-needed so your merge isn't blocked.

audreyality commented on apps/web/src/locales/en/messages.json:
🤔 The new credential generator wasn't tested in the web application, and we won't support it there until sometime after the browser extension update. Have you tested it there? If not, you definitely should.

nick-livefront replied:
I (shallowly) tested it, everything looked in order and the consuming components are behind the extension refresh flag so this would only be available after the extension refresh goes live.

audreyality commented on libs/vault/src/cipher-form/components/cipher-generator/cipher-form-generator.component.html:
```diff
@@ -1,62 +1,10 @@
-      >Placeholder: Replace with Generator Options Component(s) when available</em
-    >
-  </bit-card>
+  <tools-password-generator
+    *ngIf="type === 'password'"
+    (onGenerated)="onCredentialGenerated($event)"
+  ></tools-password-generator>
+  <tools-username-generator
+    *ngIf="type === 'username'"
+    (onGenerated)="onCredentialGenerated($event)"
+  ></tools-username-generator>
```
👍🏻 This brings joy to my life, seeing how simple the markup is. :D

audreyality commented on libs/vault/src/cipher-form/components/cipher-generator/cipher-form-generator.component.ts:
```diff
@@ -66,118 +31,8 @@ export class CipherFormGeneratorComponent implements OnChanges {
-    this.passwordTypeSubject$.next(value);
+  /** Event handler for both generation components */
+  onCredentialGenerated = (generatedCred: GeneratedCredential) => {
+    this.valueGenerated.emit(generatedCred.credential);
```
🌱 It may be relevant to send the full generated credential to the consuming component. I intend to extend this type to include metadata about the generation that we may want to store in the vault item for reporting purposes. Forwarding the full type to the consuming form keeps this component extensible. (Extending the vault item itself is out of scope.)

shane-melton replied:
I think we can leave this as is for now as we have no active or future plans to store any metadata while modifying a cipher. With that said, this is great to have available and should we need the extra information for future features its an easy enough change to make at that time.

shane-melton reviewed (dismissed):
Vault implementation looks good! Love how easy it is to use the new generator components! 🎉

One odd thing I noticed though; in your demo, the the values do not automatically regenerate in the Web client when settings are modified. However, they do in your browser extension 🤔. I don't see anything different between the client implementation on our end, so I suspect there may be something internal going awry with the new generator components.

I don't think this is blocking for this PR, but wanted to make note of it. 

cc: @danielleflinn

audreyality commented:
> Vault implementation looks good! Love how easy it is to use the new generator components! 🎉
> 
> One odd thing I noticed though; in your demo, the the values do not automatically regenerate in the Web client when settings are modified. However, they do in your browser extension 🤔. I don't see anything different between the client implementation on our end, so I suspect there may be something internal going awry with the new generator components.
> 
> I don't think this is blocking for this PR, but wanted to make note of it.
> 
> cc: @danielleflinn

The username component logic introduced a bug in the password component. I think fixing that should also fix the behavior you see in the web demo. That said, web support is out of scope of UI modernization. You may need to invoke the generator service manually in that environment if the behavior persists. (I don't think it will--the username component updates the way I'd expect it to.)

danielleflinn commented:
Agree with Shane's comment that the behavior is not blocking

A few UI nits I noticed @nick-livefront 
1. Please make sure the generator dialogs are using https://components.bitwarden.com/?path=/story/component-library-dialogs-dialog--with-cards so the background is slightly darker than the card elements
2. Is the extra spacing below the generator content in the web app coming from the `bit-section` (we have a fix to reduce this in the feature branch, but just want to make sure that is what is causing it)?

nick-livefront commented:
> 1. Please make sure the generator dialogs are using https://components.bitwarden.com/?path=/story/component-library-dialogs-dialog--with-cards so the background is slightly darker than the card elements

Fixed! [4f3e31b](https://github.com/bitwarden/clients/pull/11350/commits/4f3e31bc263adf7739a13eaba1b16b8427b5dc48)

> 2. Is the extra spacing below the generator content in the web app coming from the `bit-section` (we have a fix to reduce this in the feature branch, but just want to make sure that is what is causing it)?

Yes! Removed here: [fde20b6](https://github.com/bitwarden/clients/pull/11350/commits/fde20b69a553d1ad96452a7b5686d7aad9ab77ea)

@danielleflinn To be safe, here's screenshots of the dialog on top of the `ps/extension-refresh` branch.

| Password | Passphrase | Username |
|-|-|-|
|![Screenshot 2024-10-02 at 9 03 39 PM](https://github.com/user-attachments/assets/ece417c2-1481-4821-a1ba-b51bb5f068eb)|![Screenshot 2024-10-02 at 9 03 46 PM](https://github.com/user-attachments/assets/a4207f1a-db99-4a8f-8c3c-4a54cd34ce34)|![Screenshot 2024-10-02 at 9 03 32 PM](https://github.com/user-attachments/assets/134597d8-ec47-4986-9f97-f3e9a4e3e824)|

@audreyality While removing the margin that Danielle spotted above, I noticed that the username component had a div that had extra margin on it. This isn't needed because of the `bit-section` used above it provides margin. In this case the margin collapses and the div margin doesn't impact spacing, as shown in this video:


https://github.com/user-attachments/assets/301df2a2-319b-41dc-af39-c52935877f49

shane-melton commented on libs/vault/src/cipher-form/components/cipher-generator/cipher-form-generator.component.html:
```diff
@@ -1,4 +1,7 @@
-<bit-section>
+<div
+  disableMargin
```
❓Can we use `disableMargin` on a `div` tag? I thought it was an input property of `bit-section`.

nick-livefront replied:
😨 You're right, I missed that once I realized I no longer wanted a section component there.

[21e7ded](https://github.com/bitwarden/clients/pull/11350/commits/21e7dede79a99b71820c95ac33de72d771de1a5a)

shane-melton commented on libs/vault/src/cipher-form/components/cipher-generator/cipher-form-generator.component.html:
```diff
@@ -1,4 +1,7 @@
-<bit-section>
+<div
+  disableMargin
+  class="[&_tools-passphrase-settings_section]:tw-mb-0 [&_tools-password-settings_section]:tw-mb-0 [&_tools-username-generator_section]:tw-mb-0"
```
💭 Hmm, hopefully we can get rid of these in the future as it seems somewhat brittle if the settings sections every change or the general layout of the `tools-` generator components change.

nick-livefront replied:
Great point. Another option could be to add a prop to the two generator components to disable the trailing margin that way.

@audreyality What do you think about that?

audreyality replied:
Shane's right. You shouldn't be adjusting the inner margins of the control. So long as this effects web only, and is a minor visual artifact, it's a non-issue. Please do not modify the controls.

nick-livefront replied:
@audreyality Sounds good! Do you have an epic you're working off of or a ticket for web integration. That way I can get ahead of our QA pointing it out 😄

audreyality replied:
You can track bugs as a "QA Bug" subtask on the [layout update story](https://bitwarden.atlassian.net/browse/PM-7118). If you want more visibility than a subtask, use the parent epic.

nick-livefront replied:
🚀 [PM-13132](https://bitwarden.atlassian.net/browse/PM-13132) has been wrote to account for the spacing

audreyality commented:
> @audreyality While removing the margin that Danielle spotted above, I noticed that the username component had a div that had extra margin on it. 

That's there because the section headers can be disabled. When they don't render, the margin uncollapses to preserve spacing with whatever the parent becomes. (I don't think I've added it to the username component yet, but you can see it on the password options components.)

nick-livefront commented:
> > @audreyality While removing the margin that Danielle spotted above, I noticed that the username component had a div that had extra margin on it.
> 
> That's there because the section headers can be disabled. When they don't render, the margin uncollapses to preserve spacing with whatever the parent becomes. (I don't think I've added it to the username component yet, but you can see it on the password options components.)

👍 Good to know! This can be worked through as a part of the spacing bug as well. Thanks @audreyality !

shane-melton reviewed (approved):
Looks good, thanks! 🚀
